Content.Server/Xenoarchaeology/XenoArtifacts/ArtifactSystem.cs

```diff
@@ -1,50 +1,122 @@
+public sealed partial class ArtifactSystem : EntitySystem
 {
     [Dependency] private readonly IGameTiming _gameTiming = default!;
     [Dependency] private readonly IRobustRandom _random = default!;
-    [Dependency] private readonly IComponentFactory _componentFactory = default!;
+
+    private const int PricePerNode = 500;
+    private const int PointsPerNode = 5000;
 
     public override void Initialize()
     {
         base.Initialize();
+
         SubscribeLocalEvent<ArtifactComponent, MapInitEvent>(OnInit);
+        SubscribeLocalEvent<ArtifactComponent, PriceCalculationEvent>(GetPrice);
     }
 
     private void OnInit(EntityUid uid, ArtifactComponent component, MapInitEvent args)
     {
-        if (component.RandomTrigger)
-        {
-            AddRandomTrigger(uid, component);
-        }
+        RandomizeArtifact(component);
     }
 
-    private void AddRandomTrigger(EntityUid uid, ArtifactComponent? component = null)
+    /// <summary>
+    /// Calculates the price of an artifact based on
+    /// how many nodes have been unlocked/triggered
+    /// </summary>
+    /// <remarks>
+    /// General balancing (for fully unlocked artifacts):
+    /// Simple (1-2 Nodes): 1-2K
+    /// Medium (5-8 Nodes): 6-7K
+    /// Complex (7-12 Nodes): 10-11K
+    /// </remarks>
+    private void GetPrice(EntityUid uid, ArtifactComponent component, ref PriceCalculationEvent args)
     {
-        if (!Resolve(uid, ref component))
+        if (component.NodeTree == null)
             return;
 
-        var triggerName = _random.Pick(component.PossibleTriggers);
-        var trigger = (Component) _componentFactory.GetComponent(triggerName);
-        trigger.Owner = uid;
+        var price = component.NodeTree.AllNodes.Sum(GetNodePrice);
 
-        if (EntityManager.HasComponent(uid, trigger.GetType()))
-        {
-            Logger.Error($"Attempted to add a random artifact trigger ({triggerName}) to an entity ({ToPrettyString(uid)}), but it already has the trigger");
-            return;
-        }
+        // 25% bonus for fully exploring every node.
+        var fullyExploredBonus = component.NodeTree.AllNodes.Any(x => !x.Triggered) ? 1 : 1.25f;
+
+        args.Price =+ price * fullyExploredBonus;
+    }
+
+    private float GetNodePrice(ArtifactNode node)
+    {
+        if (!node.Discovered) //no money for undiscovered nodes.
+            return 0;
+
+        //quarter price if not triggered
+        var priceMultiplier = node.Triggered ? 1f : 0.25f;
+        //the danger is the average of node depth, effect danger, and trigger danger.
+        var nodeDanger = (node.Depth + node.Effect.TargetDepth + node.Trigger.TargetDepth) / 3;
+
+        var price = MathF.Pow(2f, nodeDanger) * PricePerNode * priceMultiplier;
+        return price;
+    }
+
+    /// <summary>
+    /// Calculates how many research points the artifact is worht
+    /// </summary>
+    /// <remarks>
+    /// Rebalance this shit at some point. Definitely OP.
```

Comment:
probably fine for this to be kind of OP since i'm not sure what the balance is going to look like between choosing to sell/destroy the artifact. i guess once the scientists get the tech tree figured out they'll hand them off to cargo instead?